### DISCUSSION TITLE
New Twenty Twenty Two Theme not yet supported by TSML

### DISCUSSION THREAD

**Discussion by kiyote33** [code4recovery/12-step-meeting-list#713] [General]
The new Block themes supported by wordpress 5.9 and newer works differently than all classic themes we are use to. Without an intervening code injection, we get a **"Deprecated: Theme without header.php is deprecated since version 3.0.0 with no alternative available. Please include a header.php template in your theme.**" error message.

**Comment by joshreisner** [comment]
interesting!

i believe this only shows up on websites that have updated their wp-config with `define('WP_DEBUG', true);` right? i don't see it with the default `false` setting.

do you think this is a problem with TSML? is there something we should change?

i see a dozen deprecation warnings on the home page with twentytwentytwo and `WP_DEBUG` turned on.

**Reply by joshreisner** [reply]
technically we could check `if (!wp_is_block_theme()) {` before the `get_header` and `get_footer` calls, that suppresses the warnings. but it also suppresses the header and the footer, which we need for layout and styling. is there another way to get header and footer output without the warning?

**Reply by kiyote33** [reply]
Hi,
I will find work arounds for this as I try to get up to speed using Gutenberg blocks.  The noted errors are only an issue with the legacy tsml interface. TSML_UI doesn't generate the same error, but I'm not having any joy yet trying to get it to display with the Twenty Twenty Two theme.

Just thought I would put this in the discussion branch so people would know we are looking into it.

**Comment by joshreisner** [comment]
I notice that the `[tsml_ui]` shortcode doesn't seem to work with the Twenty Twenty theme. I get a blank page with a console log that says `Could not find a div#tsml-ui in your HTML.`

I found that changing this line

```
wp_enqueue_script('tsml_ui', $js);
```
in shortcodes.php makes it work again. This tells it to include the script in the footer.

```
wp_enqueue_script('tsml_ui', $js, [], false, true);
```

**Reply by kiyote33** [reply]
Thx Josh,
I did find that wrapping the multiple getheader calls did work pretty well for the twenty twenty two theme.  I guess we will want to include that in a patch. 
**if( wp_is_block_theme() ) {
    block_template_part('header');
}
else {
	get_header(); 
}**

**Reply by joshreisner** [reply]
ah, nice! thanks. i created an issue

**Reply by joshreisner** [reply]
actually i spoke too soon, i can't seem to get this to work for me on my demo site. if i add this then i get no header output on twentytwentytwo. it's working for you? 

```
//do this after everything is loaded
if (wp_is_block_theme()) {
    block_template_part('header');
} else {
    get_header();
}
```

**Reply by kiyote33** [reply]
Yes, but the footer is still problematic...

**Reply by joshreisner** [reply]
`get_template_part` seems to work but still throws the same warning that `get_header` did

**Reply by joshreisner** [reply]
hm, `block_template_part` definitely does not work for me at all. i see a fully unstyled page.

![Screenshot 2022-03-04 at 08-08-58 https wordpress test](https://user-images.githubusercontent.com/1551689/156798482-1ee8abbe-94f1-4c71-9836-27e5dbb50c82.png)

**Reply by joshreisner** [reply]
i created a draft pr if you want to try it out and see if you have any ideas?
https://github.com/code4recovery/12-step-meeting-list/pull/722

**Comment by kiyote33** [comment]
FYI: We are waiting on a patch from WordPress on this, per comment from author of [WordPress Full Site Editing](https://fullsiteediting.com/lessons/how-to-use-php-templates-in-block-themes/). Discussed by TSML by-weekly team on 2022-03-11.  Currently recommend not using TwentyTwentyTwo theme until this issue resolved by the WordPress Gutenberg team.

**Reply by Grateful-Viv** [reply]
Stay away from Gutenberg 😀

**Reply by kmorning** [reply]
This is still an issue in WordPress 6.0.2, but there is a [workaround](https://github.com/WordPress/gutenberg/issues/39207#issuecomment-1087347901) posted by the same author.  Essentially, assign `do_blocks()` to variables before calling `wp_head()`.  Since the existing templates don't make a call to `wp_head()`, I added it after declaring the `do_blocks()` variables and before rendering the header.  I also found that I needed everything wrapped in a `<div class="wp-site-blocks>` for things to work properly.  Seems to work fine in my twentytwentytwo theme now.

```
...
$header = do_blocks( '<!-- wp:template-part {"slug":"header","theme":"twentytwentytwo","tagName":"header","className":"site-header","layout":{"inherit":true}} /-->' );
$footer = do_blocks( '<!-- wp:template-part {"slug":"footer","theme":"twentytwentytwo","tagName":"footer","className":"site-footer","layout":{"inherit":true}} /-->' );
?>

<head>
    <?php wp_head(); ?>
</head>

<?php
//do this after everything is loaded
if (function_exists('wp_is_block_theme') && wp_is_block_theme()) {
    echo '<div class="wp-site-blocks">';
    echo $header;
}
else {
    get_header();
}

?>
<div id="tsml">
...
...
<?php

if (function_exists('wp_is_block_theme') && wp_is_block_theme()) {
   echo $footer;
   echo '</div>';
   wp_footer();
}
else {
   get_footer();
}
?>
```
